### ISSUE TITLE
[MPS] Memory leak when converting device and type at the same time using torch.Tensor.to()

### ISSUE THREAD

**Issue by kaeru-shigure** [pytorch/pytorch#114096]
### 🐛 Describe the bug

[MPS] Memory leak when converting device and type at the same time using torch.Tensor.to()
```python
#PYTORCH_DEBUG_MPS_ALLOCATOR=1 python mps_leak_test.py
import torch

a = torch.randn(3, 1024, 1024, device='mps')

print("works fine:")
for i in range(3):
    a = a.to(device='cpu').to(dtype=torch.float32)
    a = a.to(device='mps').to(dtype=torch.float16)
    torch.mps.empty_cache()

print("memory leak:")
for i in range(3):
    a = a.to(device='cpu', dtype=torch.float32)
    a = a.to(device='mps', dtype=torch.float16)
    torch.mps.empty_cache()

print("exit")
```

```
High watermark memory allocation limit: 36.27 GB
Low watermark memory allocation limit: 29.87 GB
Initializing private heap allocator on unified device memory of size 21.33 GB
works fine:
Attempting to release cached buffers (MPS allocated: 1024.00 MB, other allocations: 14.41 MB)
Attempting to release cached buffers (MPS allocated: 1024.00 MB, other allocations: 14.41 MB)
Attempting to release cached buffers (MPS allocated: 1024.00 MB, other allocations: 14.41 MB)
memory leak:
Attempting to release cached buffers (MPS allocated: 1024.00 MB, other allocations: 26.41 MB)
Attempting to release cached buffers (MPS allocated: 1024.00 MB, other allocations: 38.41 MB)
Attempting to release cached buffers (MPS allocated: 1024.00 MB, other allocations: 50.41 MB)
exit
Attempting to release cached buffers (MPS allocated: 1024.00 MB, other allocations: 50.41 MB)
Attempting to release cached buffers (MPS allocated: 0 bytes, other allocations: 50.41 MB)
```

### Versions

PyTorch version: 2.2.0.dev20231118
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: version 3.27.7
Libc version: N/A

Python version: 3.11.5 (main, Sep 11 2023, 08:31:25) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] onnx==1.15.0
[pip3] onnxscript==0.1.0.dev20231110
[pip3] torch==2.2.0.dev20231118
[pip3] torchaudio==2.2.0.dev20231118
[pip3] torchvision==0.17.0.dev20231118
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] torch                     2.2.0.dev20231118          pypi_0    pypi
[conda] torchaudio                2.2.0.dev20231118          pypi_0    pypi
[conda] torchvision               0.17.0.dev20231118          pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr

**Comment by malfet** [comment]
Here is the fix (and I wonder how the heck it was not noticed before):
```cpp
diff --git a/aten/src/ATen/native/mps/operations/Copy.mm b/aten/src/ATen/native/mps/operations/Copy.mm
index d10b577aa72..8b1dd402e4f 100644
--- a/aten/src/ATen/native/mps/operations/Copy.mm
+++ b/aten/src/ATen/native/mps/operations/Copy.mm
@@ -140,8 +140,8 @@ static at::Tensor& copy_from_mps_(at::Tensor& dst_, const at::Tensor& src_, bool
 
       stream->copy_and_sync(
           tmpBuffer, destBuffer, size_to_copy, storage_byte_offset, destOffset, non_blocking, profile_id);
-      [destBuffer release];
     }
+    [destBuffer release];
   }
   if (!dst.is_same(dst_)) {
     dst_.copy_(dst, non_blocking);
```

**Comment by huydhn** [comment]
Confirm that this is fixed in 2.2.0 by running `python test_mps.py -v -k test_copy_cast_no_leak --repeat 10` on my M1 laptop.
